File: leetcode.com/majority-element.py
#!/bin/python3
# https://leetcode.com/problems/majority-element/
# https://www.cs.utexas.edu/~moore/best-ideas/mjrty/index.html

class Solution:
    def majorityElement(self, nums: list[int]) -> int:
        majority_num = None
        count = 0

        for num in nums:
            if count == 0:
                majority_num = num
                count += 1
            else:
                if majority_num == num:
                    count += 1
                else:
                    count -= 1
        
        return majority_num

File: leetcode.com/majority-element.rb
# @param {Integer[]} nums
# @return {Integer}
def majority_element_bf(nums)
    counter = Hash.new(0)
    nums.each { |num| counter[num] += 1 }
    k, v = counter.max_by {|k, v| v}
    k
end

def majority_element(nums)
    major = nil
    count = 0

    nums.each do |num|
        major = num if count == 0
        count += (num == major ? 1 : -1)
    end

    return major
end